docs/src/explore/design/ledger/pos-integration.md

```diff
@@ -0,0 +1,236 @@
+- deduct the slashed amount from the `validator/{validator_address}/total_deltas` at `pipeline_length` offset
+- update the `validator/{validator_address}/voting_power` for the `evidence.validator_address` in and after epoch `n + pipeline_length`
+- update the `total_voting_power` in and after epoch `n + pipeline_length`
+- update `validator_set` in and after epoch `n + pipeline_length`
+
+## Validity predicate
+
+In the following description, "pre-state" is the state prior to transaction execution and "post-state" is the state posterior to it.
+
+Any changes to PoS epoched data are checked to update the structure as described in [epoched data storage](/explore/design/pos.md#storage).
+
+Because some key changes are expected to relate to others, the VP also accumulates some values that are checked for validity after key specific logic:
+- `balance_delta: token::Change`
+- `bond_delta: HashMap<Address, token::Change>`
+- `unbond_delta: HashMap<Address, token::Change>`
+- `total_deltas: HashMap<Address, token::Change>`
+- `total_stake_by_epoch: HashMap<Epoch, HashMap<Address, token::Amount>>`
+- `expected_voting_power_by_epoch: HashMap<Epoch, HashMap<Address, VotingPower>>`: calculated from the validator's total deltas
+- `expected_total_voting_power_delta_by_epoch: HashMap<Epoch, VotingPowerDelta>`: calculated from the validator's total deltas
+- `voting_power_by_epoch: HashMap<Epoch, <HashMap<Address, VotingPower>>`
+- `validator_set_pre: Option<ValidatorSets<Address>>`
+- `validator_set_post: Option<ValidatorSets<Address>>`
+- `total_voting_power_delta_by_epoch: HashMap<Epoch, VotingPowerDelta>`
+
+The accumulators are initialized to their default values (empty hash maps and hash set). The data keyed by address are using the validator addresses.
+
+The validity predicate triggers a validation logic based on the storage keys modified by a transaction:
+
+- `validator/{validator_address}/consensus_key`:
+  ```rust,ignore
+  match (pre_state, post_state) {
+    (None, Some(post)) => {
+      // - check that any other sub-keys for this validator address didn't exist
+      // in a pre-state
+      // - check that the `state` sub-key for this validator address has been set
+      // correctly
+    },
+    (Some(pre), Some(post)) => {
+      // - check that the new consensus key is different from the old consensus
+      // key and that it has been set correctly
+    },
+    _ => false,
+  }
+  ```
+- `validator/{validator_address}/state`:
+  ```rust,ignore
+  match (pre_state, post_state) {
+    (None, Some(post)) => {
+      // - check that any other sub-keys for this validator address didn't exist
+      // in a pre-state
+      // - check that a consensus key record is also created
+      // - check that the `post` state is set correctly
+    },
+    (Some(pre), Some(post)) => {
+      // - check that a validator has been correctly deactivated or reactivated
```

Comment:
what does "correctly" mean here?

Reply:
sorry, these validation rules were a bit ambiguous here, I'll update them with more details

Reply:
updated in 9d3e1b9